src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
+        } else {
+            Node<T> current = getNodeByIndex(index);
+            Node<T> prev = current.prev;
+            Node<T> added = new Node<>(value, prev, current);
+            current.prev = added;
+            if (prev != null) {
+                prev.next = added;
+            } else {
+                first = added;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T elementFromList : list) {
+            add(elementFromList);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = getNodeByIndex(index);
+        T oldValue = currentNode.item;
+        currentNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        return unlink(nodeToRemove);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = searchNodeByItem(object);
+        if (node != null) {
+            unlink(node);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == EMPTY_SIZE;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is invalid");
+        }
+    }
+
+    private void checkIndexToAddByIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is invalid");
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        Node<T> current;
+        int middleOfSize = size / 2;
+        if (index <= middleOfSize) {
+            current = first;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = last;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private T unlink(Node<T> node) {
+        Node<T> prev = node.prev;
+        Node<T> next = node.next;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        T removedElement = node.item;
+        node.item = null;
+        size--;
+        return removedElement;
+    }
+
+    private Node<T> searchNodeByItem(T item) {
```

Comment:
Consider renaming the method 'searchNodeByItem' to 'findNodeByItem' to be consistent with naming conventions for search operations.